File: CSES/C++/Introductory Problems/Two Sets.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

vector<vector<int>> finalSubsetWithSum(vector<int>& numbers, int total) {
    int n = numbers.size();
    vector<vector<bool>> dp(n + 1, vector<bool>(total + 1, false));

    // Initialize the DP table
    for (int i = 0; i <= n; i++) {
        dp[i][0] = true;
    }

    for (int i = 1; i <= n; i++) {
        for (int j = 1; j <= total; j++) {
            dp[i][j] = dp[i - 1][j];
            if (j >= numbers[i - 1]) {
                dp[i][j] = dp[i][j] || dp[i - 1][j - numbers[i - 1]];
            }
        }
    }

    // Check if a valid subset exists
    if (!dp[n][total]) {
        return {};
    }

    // Backtrack to find the subsets
    vector<vector<int>> result;
    vector<int> currentSubset;
    int i = n, j = total;

    while (i > 0 && j > 0) {
        if (dp[i - 1][j]) {
            i--;
        } else {
            currentSubset.push_back(numbers[i - 1]);
            j -= numbers[i - 1];
            i--;
        }
    }

    result.push_back(currentSubset);
    return result;
}

int main() {
    int n;
    cin >> n;

    vector<int> mySet;
    for (int i = 1; i <= n; i++) {
        mySet.push_back(i);
    }

    int total = 0;
    for (int num : mySet) {
        total += num;
    }

    if (total % 2 == 1) {
        cout << "NO" << endl;
    } else {
        cout << "YES" << endl;
        vector<vector<int>> set1 = finalSubsetWithSum(mySet, total / 2);

        if (set1.empty()) {
            cout << "0" << endl;
        } else {
            cout << set1[0].size() << endl;
            for (int element : set1[0]) {
                cout << element << " ";
            }
            cout << endl;

            vector<int> set2;
            for (int element : mySet) {
                if (find(set1[0].begin(), set1[0].end(), element) == set1[0].end()) {
                    set2.push_back(element);
                }
            }

            cout << set2.size() << endl;
            for (int element : set2) {
                cout << element << " ";
            }
            cout << endl;
        }
    }

    return 0;
}

File: CSES/python3/Introductory Problems/Twos ets.py
n = int(input(''))

my_set = []
for i in range(1, n + 1):
    my_set.append(i)

total = sum(my_set)

def final_subset_with_sum(numbers, total):
    def subset_sum_recursive(numbers, total, index, current_subset):
        if total == 0:
            result.append(current_subset[:])
            return
        if index == len(numbers):
            return
        if total < 0:
            return
        current_subset.append(numbers[index])
        subset_sum_recursive(numbers, total - numbers[index], index + 1, current_subset)
        current_subset.pop()
        subset_sum_recursive(numbers, total, index + 1, current_subset)

    result = []
    subset_sum_recursive(numbers, total, 0, [])
    return result

if total % 2 == 1:
    print('NO')
    exit()
else:
    print('YES')
    set_1 = final_subset_with_sum(my_set, total // 2)
    print(len(set_1[0]))
    print(' '.join(map(str, set_1[0])))

    set_2 = [x for x in my_set if x not in set_1[0]]
    print(len(set_2))
    print(' '.join(map(str, set_2)))
